compiler/src/iree/compiler/Codegen/Passes.td

```diff
@@ -163,6 +163,11 @@ def GPUDistributeSharedMemoryCopy :
 def GPUPipelining : Pass<"iree-gpu-pipelining", "func::FuncOp"> {
   let summary = "Pass to do software pipelining.";
   let constructor = "mlir::iree_compiler::createGPUPipeliningPass()";
+  let options = [
+    Option<"epiloguePeeling", "epilogue-peeling", "bool",
+            /*default=*/"true",
```

Comment:
default should probably be false. Also it doesn't look like you read the option in the pass

Reply:
By default we are using epilogue peeling (the current usage in the repo), when epiloguePeeling is set to `false` we make sure prediction is available and we go ahead with un-peeled epilogue, i.e., epiloguePeeling=false.

Reply:
> By default we are using epilogue peeling (the current usage in the repo), when epiloguePeeling is set to `false` we make sure prediction is available and we go ahead with un-peeled epilogue, i.e., epiloguePeeling=false.

ah right, I read wrong. I don't see the option being used though. How does it work?

Reply:
**Peeled Epilogue**
`./tools/iree-opt --iree-gpu-pipelining /home/manigupta/cpu_machine_workspace/repos/iree/iree_tree_m1/iree/compiler/src/iree/compiler/Codegen/Common/test/gpu_pipeline.mlir`

```mlir
%18:3 = scf.for %arg0 = %c0 to %c2016 step %c32 iter_args(%arg1 = %0, %arg2 = %17, %arg3 = %c0) -> (!gpu.mma_matrix<16x16xf16, "COp">, !nvgpu.device.async.token, index) {
      nvgpu.device_async_wait %arg2 {numGroups = 0 : i32}
      gpu.barrier
      %29 = affine.apply #map6()[%2]
      %30 = gpu.subgroup_mma_load_matrix %4[%arg3, %29, %c0] {leadDimension = 40 : index} : memref<4x32x40xf16, 3> -> !gpu.mma_matrix<16x16xf16, "AOp">
      %31 = gpu.subgroup_mma_load_matrix %4[%arg3, %29, %c16] {leadDimension = 40 : index} : memref<4x32x40xf16, 3> -> !gpu.mma_matrix<16x16xf16, "AOp">
      %32 = affine.apply #map7()[%1]
      %33 = gpu.subgroup_mma_load_matrix %5[%arg3, %c0, %32] {leadDimension = 40 : index} : memref<4x32x40xf16, 3> -> !gpu.mma_matrix<16x16xf16, "BOp">
      %34 = gpu.subgroup_mma_load_matrix %5[%arg3, %c16, %32] {leadDimension = 40 : index} : memref<4x32x40xf16, 3> -> !gpu.mma_matrix<16x16xf16, "BOp">
      %35 = gpu.subgroup_mma_compute %30, %33, %arg1 : !gpu.mma_matrix<16x16xf16, "AOp">, !gpu.mma_matrix<16x16xf16, "BOp"> -> !gpu.mma_matrix<16x16xf16, "COp">
      %36 = gpu.subgroup_mma_compute %31, %34, %35 : !gpu.mma_matrix<16x16xf16, "AOp">, !gpu.mma_matrix<16x16xf16, "BOp"> -> !gpu.mma_matrix<16x16xf16, "COp">
      gpu.barrier {__pipelining_global_load__}
      %37 = arith.addi %arg0, %c32 : index
      %38 = affine.apply #map2()[%37, %1]
      %39 = affine.apply #map3()[%1, %2, %3, %workgroup_id_y]
      %40 = arith.addi %arg0, %c32 : index
      %41 = affine.apply #map8(%40)
      %42 = nvgpu.device_async_copy %6[%39, %38], %4[%41, %9, %10], 8 {__pipelining_global_load__} : memref<3456x2048xf16> to memref<4x32x40xf16, 3>
      %43 = arith.addi %arg0, %c32 : index
      %44 = affine.apply #map4()[%43, %1, %2, %3]
      %45 = affine.apply #map5()[%1, %workgroup_id_x]
      %46 = nvgpu.device_async_copy %7[%44, %45], %5[%41, %9, %10], 8 {__pipelining_global_load__} : memref<2048x1024xf16> to memref<4x32x40xf16, 3>
      %47 = nvgpu.device_async_create_group %42, %46 {__pipelining_global_load__}
      scf.yield %36, %47, %41 : !gpu.mma_matrix<16x16xf16, "COp">, !nvgpu.device.async.token, index
    }
    nvgpu.device_async_wait %18#1 {numGroups = 0 : i32}
    gpu.barrier
    %19 = affine.apply #map6()[%2]
    %20 = gpu.subgroup_mma_load_matrix %4[%18#2, %19, %c0] {leadDimension = 40 : index} : memref<4x32x40xf16, 3> -> !gpu.mma_matrix<16x16xf16, "AOp">
    %21 = gpu.subgroup_mma_load_matrix %4[%18#2, %19, %c16] {leadDimension = 40 : index} : memref<4x32x40xf16, 3> -> !gpu.mma_matrix<16x16xf16, "AOp">
    %22 = affine.apply #map7()[%1]
    %23 = gpu.subgroup_mma_load_matrix %5[%18#2, %c0, %22] {leadDimension = 40 : index} : memref<4x32x40xf16, 3> -> !gpu.mma_matrix<16x16xf16, "BOp">
    %24 = gpu.subgroup_mma_load_matrix %5[%18#2, %c16, %22] {leadDimension = 40 : index} : memref<4x32x40xf16, 3> -> !gpu.mma_matrix<16x16xf16, "BOp">
    %25 = gpu.subgroup_mma_compute %20, %23, %18#0 : !gpu.mma_matrix<16x16xf16, "AOp">, !gpu.mma_matrix<16x16xf16, "BOp"> -> !gpu.mma_matrix<16x16xf16, "COp">
    %26 = gpu.subgroup_mma_compute %21, %24, %25 : !gpu.mma_matrix<16x16xf16, "AOp">, !gpu.mma_matrix<16x16xf16, "BOp"> -> !gpu.mma_matrix<16x16xf16, "COp">
    %27 = affine.apply #map9()[%2, %workgroup_id_y]
    %28 = affine.apply #map10()[%1, %workgroup_id_x]
    gpu.subgroup_mma_store_matrix %26, %8[%27, %28] {leadDimension = 1024 : index} : !gpu.mma_matrix<16x16xf16, "COp">, memref<3456x1024xf16>
    return
  }
```

**Unpeeled Epilogue**
`./tools/iree-opt --iree-gpu-pipelining=epilogue-peeling=false /home/manigupta/cpu_machine_workspace/repos/iree/iree_tree_m1/iree/compiler/src/iree/compiler/Codegen/Common/test/gpu_pipeline.mlir`

```mlir
 %18:3 = scf.for %arg0 = %c0 to %c2048 step %c32 iter_args(%arg1 = %0, %arg2 = %17, %arg3 = %c0) -> (!gpu.mma_matrix<16x16xf16, "COp">, !nvgpu.device.async.token, index) {
      %21 = arith.cmpi slt, %arg0, %c2016 : index
      nvgpu.device_async_wait %arg2 {numGroups = 0 : i32}
      gpu.barrier
      %22 = affine.apply #map6()[%2]
      %23 = gpu.subgroup_mma_load_matrix %4[%arg3, %22, %c0] {leadDimension = 40 : index} : memref<4x32x40xf16, 3> -> !gpu.mma_matrix<16x16xf16, "AOp">
      %24 = gpu.subgroup_mma_load_matrix %4[%arg3, %22, %c16] {leadDimension = 40 : index} : memref<4x32x40xf16, 3> -> !gpu.mma_matrix<16x16xf16, "AOp">
      %25 = affine.apply #map7()[%1]
      %26 = gpu.subgroup_mma_load_matrix %5[%arg3, %c0, %25] {leadDimension = 40 : index} : memref<4x32x40xf16, 3> -> !gpu.mma_matrix<16x16xf16, "BOp">
      %27 = gpu.subgroup_mma_load_matrix %5[%arg3, %c16, %25] {leadDimension = 40 : index} : memref<4x32x40xf16, 3> -> !gpu.mma_matrix<16x16xf16, "BOp">
      %28 = gpu.subgroup_mma_compute %23, %26, %arg1 : !gpu.mma_matrix<16x16xf16, "AOp">, !gpu.mma_matrix<16x16xf16, "BOp"> -> !gpu.mma_matrix<16x16xf16, "COp">
      %29 = gpu.subgroup_mma_compute %24, %27, %28 : !gpu.mma_matrix<16x16xf16, "AOp">, !gpu.mma_matrix<16x16xf16, "BOp"> -> !gpu.mma_matrix<16x16xf16, "COp">
      gpu.barrier {__pipelining_global_load__}
      %30 = arith.addi %arg0, %c32 : index
      %31 = affine.apply #map2()[%30, %1]
      %32 = affine.apply #map3()[%1, %2, %3, %workgroup_id_y]
      %33 = arith.addi %arg0, %c32 : index
      %34 = affine.apply #map8(%33)
      %35 = arith.select %21, %c8, %c0 : index
      %36 = nvgpu.device_async_copy %6[%32, %31], %4[%34, %9, %10], 8, %35 : memref<3456x2048xf16> to memref<4x32x40xf16, 3>
      %37 = arith.addi %arg0, %c32 : index
      %38 = affine.apply #map4()[%37, %1, %2, %3]
      %39 = affine.apply #map5()[%1, %workgroup_id_x]
      %40 = arith.select %21, %c8, %c0 : index
      %41 = nvgpu.device_async_copy %7[%38, %39], %5[%34, %9, %10], 8, %40 : memref<2048x1024xf16> to memref<4x32x40xf16, 3>
      %42 = nvgpu.device_async_create_group %36, %41 {__pipelining_global_load__}
      scf.yield %29, %42, %34 : !gpu.mma_matrix<16x16xf16, "COp">, !nvgpu.device.async.token, index
    }
    %19 = affine.apply #map9()[%2, %workgroup_id_y]
    %20 = affine.apply #map10()[%1, %workgroup_id_x]
    gpu.subgroup_mma_store_matrix %18#0, %8[%19, %20] {leadDimension = 1024 : index} : !gpu.mma_matrix<16x16xf16, "COp">, memref<3456x1024xf16>
    return
```

Does it look alright?

Reply:
ah yes it does, looks like the options mechanism has changed. I get it now. This is the right solution.